File: android/core/session/src/main/java/com/anyaicliremote/core/session/SessionPayloadMapper.kt
package com.anyaicliremote.core.session

import com.anyaicliremote.core.model.ChatBlock
import com.anyaicliremote.core.model.ChatBlockKind
import com.anyaicliremote.core.model.ModelState
import com.anyaicliremote.core.model.SessionMessage
import com.anyaicliremote.core.model.SessionSummary
import com.anyaicliremote.core.model.ToolRunState
import com.anyaicliremote.core.model.WorkspaceFile
import com.anyaicliremote.core.model.ChildAgentCard
import com.anyaicliremote.core.model.childAgentStatus
import com.anyaicliremote.core.model.long
import com.anyaicliremote.core.model.obj
import com.anyaicliremote.core.model.objects
import com.anyaicliremote.core.model.string
import kotlinx.serialization.json.JsonObject

/** Pure translation of REST/RPC payloads into the UI model. */
object SessionPayloadMapper {
    fun sessions(response: JsonObject): List<SessionSummary> =
        response.objects("sessions").mapNotNull(SessionSummary::from)
            .sortedWith(compareByDescending<SessionSummary> { it.resident }.thenByDescending { it.updatedAt })

    fun historyBlocks(session: SessionSummary, messages: List<SessionMessage>): List<ChatBlock> =
        messages.mapIndexed { index, message -> historyBlock(session, index, message) }

    fun historyBlock(session: SessionSummary, index: Int, message: SessionMessage): ChatBlock {
        val blockId = "history:${session.providerId}:${session.id}:$index"
        return when (message.role) {
            "user" -> ChatBlock(blockId, ChatBlockKind.USER, text = message.content)
            "assistant" -> ChatBlock(blockId, ChatBlockKind.ASSISTANT, text = message.content)
            "tool" -> ChatBlock(
                id = blockId,
                kind = ChatBlockKind.TOOL,
                title = "工具",
                detail = message.content,
                toolState = ToolRunState.SUCCESS,
            )
            else -> ChatBlock(blockId, ChatBlockKind.SYSTEM, text = message.content)
        }
    }

    fun workspaceDirectories(response: JsonObject): List<WorkspaceFile> =
        response.objects("dirs").mapNotNull { item -> WorkspaceFile.from(item, true) }

    fun workspaceFiles(response: JsonObject): List<WorkspaceFile> =
        response.objects("files").mapNotNull { item -> WorkspaceFile.from(item, false) }

    /** Reads the history child-agent snapshot returned alongside messages. */
    fun childAgents(response: JsonObject): List<ChildAgentCard> =
        response.objects("childAgents").mapNotNull { record ->
            val providerChildId = record.string("providerChildId") ?: return@mapNotNull null
            ChildAgentCard(
                providerChildId = providerChildId,
                childSessionId = record.string("childSessionId") ?: "",
                agentType = record.string("agentType") ?: "",
                description = record.string("description") ?: "",
                status = childAgentStatus(record.string("status")),
                startedAt = record.long("startedAt"),
                completedAt = record.long("completedAt"),
                toolCallCount = record.long("toolCallCount").toInt(),
                turnCount = record.long("turnCount").toInt(),
                modelId = record.string("modelId") ?: "",
                tokensUsed = record.long("tokensUsed"),
            )
        }

    fun modelState(response: JsonObject, fallback: ModelState): ModelState {
        val source = response.obj("models")
            ?: response.obj("_meta")?.obj("modelState")
            ?: response.obj("modelState")
            ?: return fallback
        var model = fallback
        source.string("currentModelId")?.let { model = model.copy(currentModelId = it) }
        val current = source.objects("availableModels").firstOrNull { it.string("modelId") == model.currentModelId }
        val metadata = current?.obj("_meta")
        metadata?.string("reasoningEffort")?.let { model = model.copy(effort = it) }
        val levels = metadata?.objects("reasoningEfforts")?.mapNotNull { it.string("value", "id") }.orEmpty()
        if (levels.isNotEmpty()) model = model.copy(effortLevels = levels)
        return model
    }
}

File: ios/AnyAICLIRemoteCore/SessionPayloadMapper.swift
import Foundation

public enum SessionPayloadMapper {
  public struct SessionHistoryResult {
    public let session: SessionSummary
    public let blocks: [ChatBlock]
  }

  public static func sessions(from response: [String: Any]) throws -> [SessionSummary] {
    guard let rows = response["sessions"] as? [[String: Any]] else {
      throw ClientError.malformedResponse
    }
    return rows.compactMap { SessionSummary(json: $0) }.sorted {
      ($0.lastActiveAt ?? $0.createdAt ?? .distantPast)
        > ($1.lastActiveAt ?? $1.createdAt ?? .distantPast)
    }
  }

  public static func history(from response: [String: Any], fallback: SessionSummary) throws
    -> SessionHistoryResult {
    guard let metadata = response.object("session"),
      let session = SessionSummary(
        json: metadata,
        fallbackProviderID: response.string("providerId") ?? fallback.providerID
      ),
      session.id == fallback.id,
      let messages = response["messages"] as? [[String: Any]]
    else {
      throw ClientError.malformedResponse
    }
    return SessionHistoryResult(session: session, blocks: chatBlocks(from: messages))
  }

  public static func createdSession(
    from response: [String: Any],
    sessions: [SessionSummary],
    previousSessionIdentities: Set<SessionIdentity>
  ) throws -> SessionSummary? {
    let sessionID = try createdSessionID(from: response)
    let metadata = response.object("session") ?? response
    let providerID = normalized(
      metadata.string("providerId", "provider_id") ?? response.string("providerId", "provider_id"))
    let projectDirectory = normalized(
      metadata.string("projectDir") ?? response.string("projectDir"))
    if let providerID,
      let indexedSession = sessions.first(where: {
        $0.id == SessionIdentity(providerID: providerID, sessionID: sessionID)
      }) {
      return indexedSession
    }
    if providerID == nil {
      let newIdentities = Set(sessions.map(\.id)).subtracting(previousSessionIdentities)
      let matches = sessions.filter { newIdentities.contains($0.id) && $0.sessionID == sessionID }
      if matches.count == 1 { return matches[0] }
    }
    guard let providerID, let projectDirectory else { return nil }
    var normalizedMetadata = metadata
    normalizedMetadata["providerId"] = providerID
    normalizedMetadata["sessionId"] = sessionID
    normalizedMetadata["projectDir"] = projectDirectory
    return SessionSummary(json: normalizedMetadata)
  }

  public static func createdSessionID(from response: [String: Any]) throws -> String {
    guard
      let sessionID = response.string("sessionId", "session_id")
        ?? response.object("session")?.string("sessionId", "session_id")
    else {
      throw ClientError.malformedResponse
    }
    return sessionID
  }

  public static func modelState(from response: [String: Any], current: ModelState) -> ModelState {
    guard
      let source = response.object("models") ?? response.object("_meta")?.object("modelState")
        ?? response.object("modelState")
    else {
      return current
    }
    var modelState = current
    if let modelID = source.string("currentModelId") { modelState.currentModelID = modelID }
    let availableModels = source["availableModels"] as? [[String: Any]] ?? []
    if let model = availableModels.first(where: {
      $0.string("modelId") == modelState.currentModelID
    }),
      let metadata = model.object("_meta") {
      if let effort = metadata.string("reasoningEffort") { modelState.effort = effort }
      let levels = (metadata["reasoningEfforts"] as? [[String: Any]] ?? [])
        .compactMap { $0.string("value", "id") }
      if !levels.isEmpty { modelState.effortLevels = levels }
    }
    return modelState
  }

  public static func chatBlocks(from messages: [[String: Any]]) -> [ChatBlock] {
    messages.enumerated().compactMap { messageIndex, message in
      guard let role = message.string("role")?.lowercased(),
        let content = message.string("content")
      else { return nil }
      let blockID = "history-\(messageIndex)"
      switch role {
      case "system": return ChatBlock(id: blockID, kind: .system, text: content)
      case "user": return ChatBlock(id: blockID, kind: .user, text: content)
      case "assistant": return ChatBlock(id: blockID, kind: .assistant, text: content)
      case "tool":
        return ChatBlock(
          id: blockID, kind: .tool, title: "工具", detail: content, toolState: .success)
      default: return nil
      }
    }
  }

  private static func normalized(_ value: String?) -> String? {
    guard let value else { return nil }
    let trimmedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmedValue.isEmpty ? nil : trimmedValue
  }
}
